File: Prototype/c#/Prototype.cs
interface IAnimal
{
    void SetName(string name);
    
    string GetName();
    
    IAnimal Clone();
}

class Sheep : IAnimal
{
    private string _name;
    
    public Sheep() {}
    
    private Sheep(Sheep donor) => this._name = donor._name;
    
    public void SetName(string name) => this._name = name;
    
    public string GetName() => _name;
    
    public IAnimal Clone() => new Sheep(this);
}

class Program
{
    static void Main()
    {
        IAnimal sheepDonor = new Sheep();
        sheepDonor.SetName("sheep");

        IAnimal sheepClone = sheepDonor.Clone();
        
        Console.WriteLine(sheepDonor.GetName());
        Console.WriteLine(sheepClone.GetName());
    }
}

File: Prototype/java/Prototype.java
interface IAnimal{
    void setName(String name);
    
    String getName();
    
    IAnimal clone();
}

class Sheep implements IAnimal{
    private String name;
    
    public Sheep(){}
    
    private Sheep(Sheep donor){
        this.name = donor.name;
    }
    
    @Override
    public void setName(String name){
        this.name = name;
    }
    
    @Override
    public String getName(){
        return name;
    }
    
    @Override
    public IAnimal clone(){
        return new Sheep(this);
    }
}

public class Main {
    public static void main(String[] args){
        IAnimal sheepDonor = new Sheep();
        sheepDonor.setName("sheep");

        IAnimal sheepClone = sheepDonor.clone();

        System.out.println(sheepDonor.getName());
        System.out.println(sheepClone.getName());
    }
}